src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -73,7 +73,12 @@ public object GetService(Type serviceType)
             var callCount = 0;
             return provider =>
             {
-                if (Interlocked.Increment(ref callCount) == 2)
+                // PERF: Singleton services once created will always be available in the lookup dictionary.
+                // Hence it is not needed to create an lambda expression for looking up in the dictionary.
+                // This change reduces the memory allocations during the start up because 
+                // lambda expression compilation involves expensive code access security checks for Desktop CLR.
+                var singletonCallSite = callSite as SingletonCallSite;
```

Comment:
Didn't see a particularly cleaner pattern removing the polymorphism. It's used as a way to reuse code here and changing it would just involve a lot of code re-organization without any specific benefits. That said, it would make sense to move this check to `CreateServiceAccessor` and special case top level singletons outside the closure.